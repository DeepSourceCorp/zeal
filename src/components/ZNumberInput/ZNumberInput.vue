<template>
  <div class="flex items-center border rounded-md border-ink-100 focus-within:border-vanilla-400">
    <z-button
      size="x-small"
      icon="minus"
      buttonType="secondary"
      :disabled="localVal <= minNumber"
      class="flex-shrink-0"
      @click="updateNumber($event, -1)"
    />
    <input
      v-model="localVal"
      v-bind="$attrs"
      type="number"
      :min="minNumber"
      :max="maxNumber"
      :show-border="false"
      :placeholder="placeholder"
      class="flex-grow min-w-0 text-center outline-none bg-ink-400 hide-input-spinners"
      v-on="inputListeners"
    />
    <z-button
      size="x-small"
      icon="plus"
      buttonType="secondary"
      :disabled="localVal >= maxNumber"
      class="flex-shrink-0"
      @click="updateNumber($event, 1)"
    />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import ZButton from '../ZButton/ZButton.vue'

export default Vue.extend({
  name: 'ZNumberInput',
  components: { ZButton },
  props: {
    value: { type: [Number, String], default: NaN },
    min: { type: [Number, String], default: NaN },
    max: { type: [Number, String], default: NaN },
    step: { type: Number, default: 1 },
    jump: { type: Number, default: 5 },
    placeholder: { type: String, default: 'Enter a number' }
  },
  data() {
    return {
      localVal: Number(this.value)
    }
  },
  computed: {
    minNumber(): number {
      return Number(this.min)
    },
    maxNumber(): number {
      return Number(this.max)
    },
    localValNumber(): number {
      return Number(this.localVal)
    },
    inputListeners(): object {
      return Object.assign({}, this.$listeners, {
        input: ({ target }: InputEvent) => {
          if (target) {
            const a = this.minMaxNum((target as HTMLInputElement).valueAsNumber)
            this.localVal = a
            this.$emit('input', this.localValNumber)
          }
        }
      })
    }
  },
  methods: {
    updateNumber(e: MouseEvent, modifier: number): void {
      if (e.shiftKey) this.localVal = this.minMaxNum(this.localValNumber + this.jump * modifier)
      else this.localVal = this.minMaxNum(this.localValNumber + this.step * modifier)
      this.$emit('input', this.localValNumber)
    },
    minMaxNum(val: number): number {
      if (!isNaN(this.minNumber) && !isNaN(this.maxNumber))
        return Math.min(Math.max(val, this.minNumber), this.maxNumber)
      else if (!isNaN(this.minNumber)) return Math.max(val, this.minNumber)
      else if (!isNaN(this.maxNumber)) return Math.min(val, this.maxNumber)
      return val
    }
  }
})
</script>
