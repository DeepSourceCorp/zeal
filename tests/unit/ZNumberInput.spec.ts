import ZNumberInput from '../../src/components/ZNumberInput'

import { mount } from '@vue/test-utils'

const BasicTab = {
  template: `
    <div class="container space-y-5 text-vanilla-100">
      <z-number-input v-model="inputNum" class="w-28" />
      <div class="mt-2">Number: {{ inputNum }}</div>
    </div>
  `,
  components: {
    ZNumberInput
  },
  data() {
    return {
      inputNum: 0
    }
  }
}

describe('ZNumberInput', () => {
  it('renders a number input and two buttons', async () => {
    const wrapper = mount(BasicTab)

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('input').length).toEqual(1)
    expect(wrapper.findAll('button').length).toEqual(2)
  })
})
