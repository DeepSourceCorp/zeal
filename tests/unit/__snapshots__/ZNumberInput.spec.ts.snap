// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZNumberInput Decreases value by 1 on clicking \`+\` 1`] = `
<div class="container space-y-5 text-vanilla-100">
  <div class="flex items-center border rounded-md gap-x-1 border-ink-100 focus-within:border-vanilla-400 w-28"><button href="#" type="button" class="z-btn inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap flex-shrink-0 text-vanilla-100 p-0 h-6 w-6 bg-ink-300 hover:bg-ink-200"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="z-icon w-4 h-4 z-icon--minus text-vanilla-100">
        <line x1="5" y1="12" x2="19" y2="12"></line>
      </svg> </button> <input type="number" min="NaN" max="NaN" placeholder="Enter a number" class="flex-grow min-w-0 text-center outline-none bg-ink-400 hide-input-spinners"> <button href="#" type="button" class="z-btn inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap flex-shrink-0 text-vanilla-100 p-0 h-6 w-6 bg-ink-300 hover:bg-ink-200"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="z-icon w-4 h-4 z-icon--plus text-vanilla-100">
        <line x1="12" y1="5" x2="12" y2="19"></line>
        <line x1="5" y1="12" x2="19" y2="12"></line>
      </svg> </button></div>
  <div class="mt-2">Number: 0</div>
</div>
`;

exports[`ZNumberInput Increases value by 1 on clicking \`+\` 1`] = `
<div class="container space-y-5 text-vanilla-100">
  <div class="flex items-center border rounded-md gap-x-1 border-ink-100 focus-within:border-vanilla-400 w-28"><button href="#" type="button" class="z-btn inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap flex-shrink-0 text-vanilla-100 p-0 h-6 w-6 bg-ink-300 hover:bg-ink-200"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="z-icon w-4 h-4 z-icon--minus text-vanilla-100">
        <line x1="5" y1="12" x2="19" y2="12"></line>
      </svg> </button> <input type="number" min="NaN" max="NaN" placeholder="Enter a number" class="flex-grow min-w-0 text-center outline-none bg-ink-400 hide-input-spinners"> <button href="#" type="button" class="z-btn inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap flex-shrink-0 text-vanilla-100 p-0 h-6 w-6 bg-ink-300 hover:bg-ink-200"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="z-icon w-4 h-4 z-icon--plus text-vanilla-100">
        <line x1="12" y1="5" x2="12" y2="19"></line>
        <line x1="5" y1="12" x2="19" y2="12"></line>
      </svg> </button></div>
  <div class="mt-2">Number: 0</div>
</div>
`;

exports[`ZNumberInput renders a number input and two buttons 1`] = `
<div class="container space-y-5 text-vanilla-100">
  <div class="flex items-center border rounded-md gap-x-1 border-ink-100 focus-within:border-vanilla-400 w-28"><button href="#" type="button" class="z-btn inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap flex-shrink-0 text-vanilla-100 p-0 h-6 w-6 bg-ink-300 hover:bg-ink-200"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="z-icon w-4 h-4 z-icon--minus text-vanilla-100">
        <line x1="5" y1="12" x2="19" y2="12"></line>
      </svg> </button> <input type="number" min="NaN" max="NaN" placeholder="Enter a number" class="flex-grow min-w-0 text-center outline-none bg-ink-400 hide-input-spinners"> <button href="#" type="button" class="z-btn inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap flex-shrink-0 text-vanilla-100 p-0 h-6 w-6 bg-ink-300 hover:bg-ink-200"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="z-icon w-4 h-4 z-icon--plus text-vanilla-100">
        <line x1="12" y1="5" x2="12" y2="19"></line>
        <line x1="5" y1="12" x2="19" y2="12"></line>
      </svg> </button></div>
  <div class="mt-2">Number: 0</div>
</div>
`;
